### PR TITLE
Hazard Test Plan

### PR DESCRIPTION
**Pull request by davidharrishmc**

### CODE REVIEW

File: docs/ctp/src/unpriv.adoc [hunk 1]
```diff
@@ -189,6 +189,72 @@ Notes:
 The register edge cases are selected to include extreme values and their immediate neighbors.  They also include walking 1s and an arbitrary intermediate value.  RV64 adds edge cases near the 32-bit boundary to stress W-type instructions.  Immediate edge cases exercise each bit of the immediate, extreme values, and one random intermediate value.
 ====
 
+==== Hazards
+
+Hazard tests are applied to every unprivileged _Instruction Under Test_ (IUT) that uses a source or destination.
+The hazard receives the source or sends the destination value to a _Secondary Instruction_ (SI).  All relevant
```
Comment (jordancarlin) [inline docs/ctp/src/unpriv.adoc]:
```suggestion
The hazard receives the source from or sends the destination value to a _Secondary Instruction_ (SI).  All relevant
```

File: docs/ctp/src/unpriv.adoc [hunk 2]
```diff
@@ -189,6 +189,72 @@ Notes:
...
+instruction from each of many categories of instructions to exercise forwarding between different functional units.
+
+The categories of SI are given in <<t-secondary-instruction-categories>>.
+
+[[t-secondary-instruction-categories]]
+.Secondary Instruction Categories
+[options=header]
+|===
+|Category|ALU|Memory|Control|CSR|Mult|Float|VectorInt|VectorFloat
+|Extension||||Zicsr|M|F|Zve32x, Zvl64b|Zve32f, Zvl64b
```
Comment (jordancarlin) [inline docs/ctp/src/unpriv.adoc]: How about folding the "Extension" row into the Category row. Or changing "Extension" to something that makes it clear the the column only applies if the extension is implemented.

File: docs/ctp/src/unpriv.adoc [hunk 3]
```diff
@@ -189,6 +189,72 @@ Notes:
...
+|Int Dest (rd)|auipc, li|lw, amoswap.w (Zaamo)|jal|csrr|mul, div|fcvt.w.s, fmv.x.w, fclass.s, feq.s|vmv.x.s|
+|Int Source|add (rs1/2), addi (rs2)|sw (rs2), amoswap.w (rs2, Zaamo)|csrw (rs1)|beq (rs1/2)|mul (rs1/2), div (rs1/2)|fcvt.s.w (rs1), fmv.w.x (rs1)
```
Comment (jordancarlin) [inline docs/ctp/src/unpriv.adoc]: `mul` could come from just Zmmul but `div` needs the full M extension.

File: docs/ctp/src/unpriv.adoc [hunk 4]
```diff
@@ -189,6 +189,72 @@ Notes:
...
+
+The test cases are as follows:
+
+* If the IUT has {int/fp/vector/memory/CSR} source operand:
+** For each IUT source
+*** If IUT has a destination, hazard_types = {RAW, WAW}.  Else hazard_types = RAW.
+*** For each hazard_type
+**** For each supported SI category that can write the source type (gated by #ifdef EXT_SUPPORTED)
+***** For each delay in (0, random[1,3])
+****** Pick random unused registers for IUT source(s) and destination from X1-x15, f0-f31, v0-v31.  Ok to duplicate source registers.
```
Comment (jordancarlin) [inline docs/ctp/src/unpriv.adoc]: No need to constrain to x1-x15. Hazard coverage will be in unpriv tests, which already generates separate tests for I vs E.

File: docs/ctp/src/unpriv.adoc [hunk 5]
```diff
@@ -189,6 +189,72 @@ Notes:
...
+****** Pick random representative SI from category, and random unused source registers for SI.
+****** Initialize IUT and SI sources to random values
+****** If hazard_type = RAW, SI dest = random IUT source
+****** If hazard_type = WAW, SI dest = IUT dest
+****** Perform SI
+****** If delay > 0, perform delay dependent instructions
```
Comment (jordancarlin) [inline docs/ctp/src/unpriv.adoc:294]: Should we specify what the delay dependent instruction is?

File: docs/ctp/src/unpriv.adoc [hunk 6]
```diff
@@ -2,7 +2,7 @@
 [appendix]
 == Unprivileged Test Plan
 
-Unprivileged tests exercise every instruction using every applicable source and destination register, and reasonable architectural edge values of sources.  They are intended for certification, not verification. For example, they do not stress pipeline hazards.
+Unprivileged tests exercise every instruction using every applicable source and destination register, and reasonable architectural edge values of sources.  They are intended for certification, not verification. For example, they do not stress interrupts arriving in the midst of multicycle instructions.
```
[the remark this answers is not in the document]
Comment (davidharrishmc) [reply docs/ctp/src/unpriv.adoc:5]: Copilot is confused.

[the remark this answers is not in the document]
Comment (jordancarlin) [reply docs/ctp/src/unpriv.adoc:5]: Very confused. This doesn't make any sense.

File: docs/ctp/src/unpriv.adoc (hunk 5 above)
Comment (davidharrishmc) [reply docs/ctp/src/unpriv.adoc:294]: Leave it to you.

**Review by jordancarlin — APPROVED** [review]

File: docs/ctp/src/unpriv.adoc [hunk 7]
```diff
@@ -189,6 +189,71 @@ Notes:
...
+with the SI being adjacent to the IUT, and being separated by a small amount.  The SI is selected as a representative
+instruction from each of many categories of instructions to exercise forwarding between different functional units.
+
+The categories of SI are given in <<t-secondary-instruction-categories>>.
+
+[[t-secondary-instruction-categories]]
+.Secondary Instruction Categories
+[options=header]
+|===
+|Category|ALU|Memory|Control|CSR (Zicxr)|Mult (Zmmul/M)|Float (F)|VectorInt (Zve32x + Zvl64b)|VectorFloat (Zve32f+Zvl64b)
```
Comment (qbojj) [inline docs/ctp/src/unpriv.adoc]: Shouldn't this be Zicsr rather than Zicxr?

Comment (davidharrishmc) [reply docs/ctp/src/unpriv.adoc]: Thx!
